File: wave.py
import numpy
from scipy.special import hankel1

ZHAT = numpy.array([0., 0., 1.])
PI = numpy.pi
TWOPI = 2. * PI
FOURPI = 2. * TWOPI


def incident(kvec, points):
	"""
	Incident wave

	@param kvec incident wave vector
	@param point target points, of size n * 3
	@return complex number
	"""
	n = points.shape[0]
	return numpy.exp([1j*kvec.dot(points[i, :]) for i in range(n)])


def gradIncident(kvec, nDotK, points):
	"""
	Normal gradient of the incident wave, assumes incident wave is exp(1j * kvec.x)

	@param kvec incident wave vector
	@param nDotK nvec . kvec array of size n
	@param points (source) point, array of size n * 3
	@return complex number
	"""
	return 1j*nDotK*incident(kvec, points)


def computeScatteredWave(kvec, xc, yc, point):
	"""
	Total scattered wave response, summing up 
	contributions from each segment

	@param kvec incident wave vector
	@param xc list of x coordinates representing the contour, must close
	@param yc list of y coordinates representing the contour, must close
	@param point observer point
	@return complex value
	"""

	kmod = numpy.sqrt(kvec.dot(kvec))

	n = len(xc)
	nm1 = n - 1
	pc = numpy.array([(xc[i], yc[i], 0.) for i in range(n)])
	xdot = pc[1:, :] - pc[:-1, :]
	pmid = 0.5*(pc[1:, :] + pc[:-1, :])
	dsdt = numpy.sqrt(xdot[:, 0]*xdot[:, 0] + xdot[:, 1]*xdot[:, 1])
	nvec = numpy.array([numpy.cross(ZHAT, xdot[i, :]) for i in range(nm1)])
	rvec = numpy.array([point - pmid[i, :] for i in range(nm1)])
	r = numpy.sqrt(rvec[:, 0]*rvec[:, 0] + rvec[:, 1]*rvec[:, 1])
	kr = kmod * r
	nDotR = numpy.array([nvec[i, :].dot(rvec[i, :]) for i in range(nm1)])
	nDotK = numpy.array([nvec[i, :].dot(kvec) for i in range(nm1)])

	g = (1j/4.) * hankel1(0, kr)
	dgdn = (-1j/4.) * hankel1(1, kr) * kmod * nDotR / r

	# contribution from the gradient of the incident wave on the surface
	# of the obstacle. The normal derivative of the scattered wave is 
	# - normal derivative of the incident wave.
	scattered_wave = - dsdt * g * gradIncident(kvec, nDotK, pmid)

	# shadow side: total wave is nearly zero 
	#              => scattered wave amplitude = -incident wave ampl.
	#
	# illuminated side:
	#              => scattered wave amplitude = +incident wave ampl.
	shadow = (nDotK > 0.) - 0.5
	scattered_wave += shadow * dsdt * dgdn * incident(kvec, pmid)

	return scattered_wave.sum()

File: src/wave.cpp
#include "wave.h"
#include <boost/math/special_functions/bessel.hpp>

std::complex<double>
hankel1(int n, double x) {
	double besselJ = boost::math::cyl_bessel_j<int, double>(n, x);
	double besselY = boost::math::cyl_neumann<int, double>(n, x);
	return besselJ + J1*besselY;
}

std::complex<double> 
incident(const double kvec[], const double point[]) {
	return std::exp(J1*(kvec[0]*point[0] + kvec[1]*point[1]));
}


std::complex<double> 
gradIncident(const double nvec[], const double kvec[], 
	         const double point[]) {
	return J1*(nvec[0]*kvec[0] + nvec[1]*kvec[1])*incident(kvec, point);
}

std::complex<double> 
computeScatteredWaveElement(const double kvec[], const double p0[], 
	                        const double p1[], const double point[]) {

    // xdot is anticlockwise
    double xdot[] = {p1[0] - p0[0], p1[1] - p0[1]};

    // mid point of the segment
    double pmid[] = {0.5*(p0[0] + p1[0]), 0.5*(p0[1] + p1[1])};

    // segment length
    double dsdt = sqrt(xdot[0]*xdot[0] + xdot[1]*xdot[1]);

    // normal vector, pointintg inwards and normalised
    double nvec[] = {-xdot[1]/dsdt, xdot[0]/dsdt, 0.};

    // from segment mid-point to observer
    double rvec[] = {point[0] - pmid[0], point[1] - pmid[1]};
    double r = sqrt(rvec[0]*rvec[0] + rvec[1]*rvec[1]);

    double kmod = sqrt(kvec[0]*kvec[0] + kvec[1]*kvec[1]);
    double kr = kmod * r;

    // Green functions and normal derivatives
    std::complex<double> g = (J1/4.) * hankel1(0, kr);
    double nDotR = nvec[0]*rvec[0] + nvec[1]*rvec[1];
    std::complex<double> dgdn = (-J1/4.) * hankel1(1, kr) * kmod * nDotR / r;

    // contribution from the gradient of the incident wave on the surface
    // of the obstacle. The normal derivative of the scattered wave is
    // - normal derivative of the incident wave.
    std::complex<double> scattered_wave = - dsdt * g * gradIncident(nvec, kvec, pmid);

    // shadow side: total wave is nearly zero
    //              => scattered wave amplitude = -incident wave ampl.
    //
    // illuminated side:
    //              => scattered wave amplitude = +incident wave ampl.
    //
    double nDotK = nvec[0]*kvec[0] + nvec[1]*kvec[1];          
    double shadow = (nDotK > 0 ? 1.0 : -1.0);
        
    scattered_wave += shadow * dsdt * dgdn * incident(kvec, pmid);

    return scattered_wave;
}

std::complex<double> 
computeScatteredWave(const double kvec[], int nc, const double xc[], const double yc[], 
	                 const double point[]) {
	double p0[2], p1[2];
	std::complex<double> res(0., 0.);
	for (int i = 0; i < nc - 1; ++i) {
		p0[0] = xc[i]; p0[1] = yc[i];
		p1[0] = xc[i + 1]; p1[1] = yc[i + 1];
		res += computeScatteredWaveElement(kvec, p0, p1, point);
	}
	return res;
}

